File: nodes/subscriber.py
#!/usr/bin/env python3

from collections import defaultdict
import rclpy
from rclpy.node import Node
from rclpy.serialization import serialize_message

from sensor_msgs.msg import PointCloud2
from sensor_msgs.msg import Image
from tf2_msgs.msg import TFMessage
from sensor_msgs.msg import CameraInfo

from collections import defaultdict
import numpy as np
import sys
import time

BUFFER_SIZE = 100
THROUGHPUT = 0

def get_statistics(msg_info, msg_size, topic):
    global THROUGHPUT
    msg_info = np.array(msg_info)
    # Ignore the first entry, as we serialize the msg to get its size when the callback is run the first time.
    # This makes the callback run much longer.
    latencies = msg_info[1:,0]
    current_throughput = round((msg_size * len(latencies))/((msg_info[-1,1] - msg_info[1,1]) * 1024 *1024), 3)
    THROUGHPUT += current_throughput

    report = "\n------------------" +\
    "\nFor topic " + topic +\
    "\nLatency mean(ms): " + str(round(np.mean(latencies) * 1000, 3)) +\
    "\nLatency minimum(ms): " + str(round(min(latencies) * 1000, 3)) +\
    "\nLatency maximum(ms): " + str(round(max(latencies) * 1000, 3)) +\
    "\nLatency std dev(ms): " + str(round(np.std(latencies) * 1000, 3)) +\
    "\nMsg size (bytes): " + str(msg_size) +\
    "\nThroughput (MB/sec): " + str(current_throughput) +\
    "\nCumulative throughput (MB/sec): " + str(THROUGHPUT) +\
    "\n------------------"
    return report

class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__('subscriber_py')

        self.data_buffer = defaultdict(list)
        self.msg_sizes = defaultdict(int)
        topics_and_types = [('/k4a/points', PointCloud2), ('/k4a/color/image', Image), ('/k4a/depth/image', Image),
                            ('/k4a/color/camera_info', CameraInfo), ('/k4a/depth/camera_info', CameraInfo)]
        self.subscribers = []

        for topic, type in topics_and_types :
            self.subscribers.append(self.create_subscription(
                type,
                topic,
                self.msg_callback_generator(topic),
                10))

    def msg_callback_generator(self, topic):
        def msg_callback(msg):
            if len(self.data_buffer[topic]) == BUFFER_SIZE:
                return

            src_time = float(str(msg.header.stamp.sec) + "." + str(msg.header.stamp.nanosec))
            current_time = float(time.time())
            latency = current_time - src_time

            if self.msg_sizes[topic] == 0:
                bytes = serialize_message(msg)
                self.msg_sizes[topic] = sys.getsizeof(bytes)

            if latency > 0 and len(self.data_buffer[topic]) < BUFFER_SIZE:
                self.data_buffer[topic].append([latency, current_time])
 
            if latency < 0 :
                self.get_logger().info("Timing mismatch in " + topic + " , dropping msg")

            if len(self.data_buffer[topic]) == BUFFER_SIZE :
                self.get_logger().info(get_statistics((self.data_buffer[topic]), self.msg_sizes[topic] , topic))

        return msg_callback

def main(args=None):
    rclpy.init(args=args)

    minimal_subscriber = MinimalSubscriber()

    rclpy.spin(minimal_subscriber)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    minimal_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: nodes/subscriber_cpp/subscriber.cpp
#include <memory>
#include <string>
#include <vector>
#include <map>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/serialization.hpp"
#include "rclcpp/serialized_message.hpp"

#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "sensor_msgs/msg/camera_info.hpp"

long unsigned int BUFFER_SIZE = 101;
double THROUGHPUT = 0;

std::map<std::string, int> msg_sizes {
  {"/k4a/points", 0},
  {"/k4a/depth/image", 0},
  {"/k4a/color/image", 0},
  {"/k4a/depth/camera_info", 0},
  {"/k4a/color/camera_info", 0},
};

std::map<std::string, std::vector<double>> current_timestamps {
  {"/k4a/points", {}},
  {"/k4a/depth/image", {}},
  {"/k4a/color/image", {}},
  {"/k4a/depth/camera_info", {}},
  {"/k4a/color/camera_info", {}},
};

template <typename T>
void msg_callback(const T msg, std::string topic, rclcpp::Time current_time) {

  if (current_timestamps[topic].size() == BUFFER_SIZE) {
    return;
  }

  auto logger = rclcpp::get_logger("subscriber_cpp");

  double current_nsec = current_time.nanoseconds();

  if (current_timestamps[topic].size() < BUFFER_SIZE) {
    current_timestamps[topic].push_back(current_nsec);
  }

  if (current_timestamps[topic].size() == BUFFER_SIZE) {
    int n = current_timestamps[topic].size();
    double data_flow = msg_sizes[topic] * (n) / (1024 * 1024);
    double delta_t = (current_timestamps[topic][n-1] - current_timestamps[topic][0])/ 1000000000 ;
    auto throughput = data_flow/delta_t;

    THROUGHPUT += throughput;

    std::string report =
      std::string("\n--------------") +
      std::string("\nFor topic: ") + topic +
      std::string("\nMsg size (bytes): ") + std::to_string(msg_sizes[topic]) +
      std::string("\nThroughput (MB/s): ") + std::to_string(throughput) +
      std::string("\nCumulative throughput (MB/s): ") + std::to_string(THROUGHPUT) +
      std::string("\n--------------");
    
      RCLCPP_INFO(logger, report.c_str());
  }

  // Find the size of the msg
  if (msg_sizes[topic] == 0) {
    rclcpp::Serialization<T> serializer;
    rclcpp::SerializedMessage serialized_msg;
    serializer.serialize_message(&msg, &serialized_msg);
    msg_sizes[topic] = serialized_msg.size();
  }
}

class MinimalSubscriber : public rclcpp::Node
{
  public:
    // Subscriber objects
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr subscription_pointcloud;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_depth;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_color;
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr subscription_color_caminfo;
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr subscription_depth_caminfo;

    MinimalSubscriber()
    : Node("subscriber_cpp")
    {
      // Start subscribers
      subscription_pointcloud = this->create_subscription<sensor_msgs::msg::PointCloud2>(
      "/k4a/points", 10,
      [&](sensor_msgs::msg::PointCloud2 msg) {
        msg_callback<sensor_msgs::msg::PointCloud2>(msg, "/k4a/points", this->get_clock()->now());
      });

      subscription_depth = this->create_subscription<sensor_msgs::msg::Image>(
      "/k4a/depth/image", 10,
      [&](sensor_msgs::msg::Image msg) {
        msg_callback<sensor_msgs::msg::Image>(msg, "/k4a/depth/image", this->get_clock()->now());
      });

      subscription_color = this->create_subscription<sensor_msgs::msg::Image>(
      "/k4a/color/image", 10,
      [&](sensor_msgs::msg::Image msg) {
        msg_callback<sensor_msgs::msg::Image>(msg, "/k4a/color/image", this->get_clock()->now());
      });

      subscription_depth_caminfo = this->create_subscription<sensor_msgs::msg::CameraInfo>(
      "/k4a/depth/camera_info", 10,
      [&](sensor_msgs::msg::CameraInfo msg) {
        msg_callback<sensor_msgs::msg::CameraInfo>(msg, "/k4a/depth/camera_info", this->get_clock()->now());
      });

      subscription_color_caminfo = this->create_subscription<sensor_msgs::msg::CameraInfo>(
      "/k4a/color/camera_info", 10,
      [&](sensor_msgs::msg::CameraInfo msg) {
        msg_callback<sensor_msgs::msg::CameraInfo>(msg, "/k4a/color/camera_info", this->get_clock()->now());
      });
    }
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalSubscriber>());
  rclcpp::shutdown();
  return 0;
}
